File: javascript/robot-simulator/robot-simulator.js
export class InvalidInputError extends Error {
  constructor(...args) {
    super(...args);
  }
}

export class Robot {
  constructor() {
    this.moves = {
      east: { x: 1, y: 0, left: 'north', right: 'south' },
      west: { x: -1, y: 0, left: 'south', right: 'north' },
      north: { x: 0, y: 1, left: 'west', right: 'east' },
      south: { x: 0, y: -1, left: 'east', right: 'west' },
    };
    this.position = { x: 0, y: 0 };
    this.direction = 'north';
  }

  get bearing() {
    return this.direction;
  }

  get coordinates() {
    return [this.position.x, this.position.y];
  }

  at(x, y) {
    this.place({ x, y, direction: this.direction });
  }

  orient(direction) {
    this.place({ x: this.position.x, y: this.position.y, direction });
  }

  place({ x, y, direction }) {
    if (!(direction in this.moves)) {
      throw new InvalidInputError(`Invalid direction: ${direction}`);
    }
    this.position = { x, y };
    this.direction = direction;
  }

  advance() {
    const move = this.moves[this.direction];
    this.position = { x: this.position.x + move.x, y: this.position.y + move.y };
  }

  turnRight() {
    this.direction = this.moves[this.direction].right;
  }

  turnLeft() {
    this.direction = this.moves[this.direction].left;
  }

  evaluate(input) {
    Robot.instructions(input).forEach((method) => this[method]());
  }

  static instructions(input) {
    return [...input].map((instruction) => {
      if (instruction === 'A') {
        return 'advance';
      } else if (instruction === 'R') {
        return 'turnRight';
      } else if (instruction === 'L') {
        return 'turnLeft';
      } else {
        throw new InvalidInputError(`Invalid instruction: ${instruction}`);
      }
    });
  }
}

File: python/robot-simulator/robot_simulator.py
EAST = complex(1, 0)
NORTH = complex(0, 1)
WEST = complex(-1, 0)
SOUTH = complex(0, -1)

class Robot:
    def __init__(self, direction=NORTH, x_pos=0, y_pos=0):
        self.position = complex(x_pos, y_pos)
        self.direction = direction

        self._moves = {
            'L': self.turn_left,
            'R': self.turn_right,
            'A': self.advance,
        }

    @property
    def coordinates(self):
        return self.position.real, self.position.imag

    def turn_left(self):
        self.direction *= 1j

    def turn_right(self):
        self.direction *= -1j

    def advance(self):
        self.position += self.direction

    def move(self, instructions):
        for instruction in instructions:
            self._moves[instruction]()
